🚀 2단계 - 리다이렉트 필터, Oauth 인증 필터 리뷰 요청드립니다.

parkjun5 opened the pull request:
안녕하세요.
🚀2-1단계 - 리다이렉트 필터
🚀2-2단계 - Oauth 인증 필터 리뷰 요청드립니다.

이번 미션은 이전 미션에서 작동한 필터에 provider 체크, OAuth2LoginAuthenticationFilter 가 이전 필터를 제대로 체크를 추가하는 것을 깔끔하게 중요했던 것 같습니다.
또한 OAuth2ClientRegistration와 OAuth2AuthorizationRequest 가 추가되면서 Clinet(서버) 의 정보를 핸들링하는 부분이 많이 변경되었습니다.

저는 OAuth2ClientRegistration는 서버 정보를 갖도록 하고 OAuth2AuthorizationRequest 에서 시큐리티에서 필요한 정보만 남도록 하였는데, ProviderDetails 와 UserInfoEndpoint는 마지막까지 사용되지 않았네요.

아마 여러 인증 요청을 state 값으로 구분하다보니 생긴 문제 같네요. state 를 구분 값으로  사용한다는 말이 있어서 추가하였는데, 결과물을 보니 잘못 사용하고 있는 것 같네요 ㅎㅎ;;

많은 피드백 부탁 드립니다!

parkjun5 committed:
ref: change direct session inserts to save by SecurityContextRepository

parkjun5 committed:
ref: strategies should be toUnmodifiableMap

 - Use UriComponentsBuilder

parkjun5 committed:
ref: Add OAuth2ProviderSupportChecker for safety

parkjun5 committed:
feat: add OAuth2 authorization request handling

- Add ClientRegistration 검증 github or google 만 가능하도록 변경
- OAuth2ProviderSupportChecker 체크를 ClientRegistration 로 변경
- Add AuthorizationRequest 저장 로직 추가 이를 통해 Client 설정 파일을 한곳에서 관리하도록 변경
- InmemoryClientRegistrationRepository 는 빈 초기화시 최초로 서버에서 제공하는 ClientRegistration 를 갖는다.

parkjun5 committed:
polishing

parkjun5 committed:
ref: OAuth2AuthenticationRequestResolver 가 단순 인증 URI을 주지 않고 OAuth2AuthorizationRequest를 만들어준다.

 - 서버에 저장된 OAuth2ClientRegistration를 기준으로 OAuth2AuthorizationRequest 를 만들어준다.
 - OAuth2AuthorizationRequest 의 구분을 위해 state 추가 UUID를 통해 기본 값을 설정해준다.

parkjun5 committed:
ref: TokenRequestStrategy 에서는 더이상 클라이언트 정보를 가지지 않는다.

parkjun5 committed:
ref: OAuth2UserService 를 통해 서버에 저장된 OAuth2 요청 정보와 ClientRegistration 정보를 가져온다.

parkjun5 committed:
ref: OAuth2AuthorizationRequest 를 저장할 OAuth2AuthorizationRecord 추가

parkjun5 committed:
ref: 초반 사용하던 oAuth2Type를 registrationId로 변경

parkjun5 committed:
ref: OAuth2LoginAuthenticationFilter 는 OAuth2AuthorizationRequest를 통해 인가를 진행하도록 변경

 - OAuth2AuthorizationRequest 가 존재하지 않으면 진행할 수 없다.
 - OAuth2AuthorizationRequest 를 통해 서버 내부 값을 가져온다.

parkjun5 committed:
test: state 추가로 인한 테스트 수정

parkjun5 committed:
ref: OAuth2AuthorizationRecord, OAuth2AuthorizationRequest 과도하게 많은 값을 가지고 있던 문제 해결

 - 특히 ClientSecret 등 민감한 정보와 많은 값을 가지고 있던 문제 해결

parkjun5 committed:
ref: 서버의 속성값은 registrationId를 통해 저장된 clientRegistration를 사용하도록 변경

parkjun5 committed:
ref: OAuth2Type 이라고 사용하던 제공하는 OAuth2 서비스의 이름을 전부 RegistrationId 로 변경

parkjun5 committed:
docs : Add READ_ME.md

parkjun5 committed:
ref : OAuth2LoginAuthenticationFilter 권장 사항에 맞춰 정리

parkjun5 committed:
polishing

parkjun5 committed:
ref: remove OAuth2 configuration from Strategy pattern

- OAuth2 설정들을 Properties 빈으로 이동
- Strategy 인터페이스에는 동작에 필요한 값만 가지도록 변경

parkeeseul commented on src/main/java/nextstep/app/application/OAuth2UserServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.NoSuchElementException;
+
+@Service
+public class OAuth2UserServiceImpl implements OAuth2UserService {
```
Spring Security에서 OAuth2UserService는 OAuth2 인증 후 사용자 정보를 로드하는 역할을 합니다.
하지만 현재 OAuth2UserServiceImpl는 클라이언트 등록 정보(ClientRegistration) 및 인증 요청(OAuth2AuthorizationRequest)을 관리하는 서비스로 사용하고 있네요 

spring security의 DefaultOAuth2UserService 를 한번 참고하여 어떤 연할을 하고 있는지도 학습해보면 좋을 것 같습니다 😄

parkeeseul commented on src/main/java/nextstep/security/authentication/filter/oauth/OAuth2RedirectAuthenticationFilter.java:
```diff
@@ -2,60 +2,52 @@
+            return;
+        }
 
-            if (noRequiresAuthentication(request)) {
-                filterChain.doFilter(request, response);
-                return;
-            }
+        String registrationId = getRegistrationId(request.getRequestURI());
+        OAuth2ClientRegistration clientRegistration = auth2UserService.loadClientRegistrationByRegistrationId(registrationId);
 
-            String oAuth2Type = getOAuth2Type(request.getRequestURI());
+        OAuth2AuthorizationRequest authorizationRequest = oAuth2AuthenticationRequestResolver.resolve(clientRegistration);
```
현재 코드에서는 `registrationId`를 직접 추출하고, `OAuth2ClientRegistration`을 조회한 후,  
해당 정보를 사용하여 `OAuth2AuthorizationRequest`를 생성하고 있는데요,  

Spring Security의 `OAuth2AuthorizationRequestResolver`는  
`resolve(HttpServletRequest request)`를 통해  
이러한 과정(`registrationId` 추출 → `ClientRegistration` 조회 → `OAuth2AuthorizationRequest` 생성)을 **내부적으로 처리**하고 있습니다.  

즉, `OAuth2AuthorizationRequestResolver`에 위임하면  
 **불필요한 로직을 제거하여 코드가 더 간결해지고**,  
 **OAuth2 요청 처리를 한 곳에서 관리할 수 있어 책임이 분리**됩니다.  

아래와 같이 `OAuth2AuthorizationRequestResolver`에 위임하면 더 간결한 코드가 될 것 같은데,  
한번 고민해 보셔도 좋을 것 같습니다! 😊  

```java
OAuth2AuthorizationRequest authorizationRequest = oAuth2AuthenticationRequestResolver.resolve(request);
```

parkjun5 replied:
필터에서 굳이 생성되는 모든 객체를 끄집어 내서 알 필요는 없겠네요! 리졸버가 해당 내용을 갖도록 변경해보겠습니다.

parkeeseul commented on src/main/java/nextstep/security/userservice/OAuth2ClientRegistration.java:
```diff
@@ -0,0 +1,9 @@
+package nextstep.security.userservice;
+
+
+public interface OAuth2ClientRegistration {
```
해당 interface는 어떠한 목적으로 설계해 주셨나요 ? 😄

parkjun5 replied:
application 과 security를 분리한다는 느낌으로 인터페이스로 만들어 보았어요. 
클래스도 좋지만 UserDetail 처럼 다른 라이브러리 처럼 사용하는게 어떤가 해서요. 
정리하면 확장성과 관심사 분리를  위해서 인터페이스로 설계하였습니다.

parkeeseul replied:
고민하신 부분을 공유해 주셔서 감사합니다! 😊  

제가 생각하기에 OAuth2 인증 설정 객체는 비즈니스 도메인 로직과 직접적인 관련이 없고,  
오히려 `security` 패키지에서 관리하는 것이 더 적절하다고 생각합니다.  

실제로 Spring Security에서도 `ClientRegistration`을 `security.oauth2.client.registration` 패키지에서 관리하고 있어,  
보안 관련 설정을 도메인 로직과 분리하는 것이 일반적인 패턴이라고 생각합니다. 😄

parkeeseul commented on src/main/java/nextstep/app/domain/ClientRegistration.java:
```diff
@@ -0,0 +1,79 @@
+        String clientId,
+        String clientSecret,
+        String redirectUri,
+        String scope
```
scope는 여러 개의 권한이 포함될 수 있어 Set<String>으로 관리해도 좋을 것 같네요 😄

parkjun5 replied:
우선 현재 기능에서는 고정된 scope 만 사용하고 있어서 String으로 두었는데, 바꿔보겠습니다.

parkeeseul commented on src/main/java/nextstep/security/authentication/oauth/OAuth2AuthorizationRequest.java:
```diff
@@ -0,0 +1,9 @@
+package nextstep.security.authentication.oauth;
+
+public interface OAuth2AuthorizationRequest {
```
OAuth2AuthorizationRequest 를 interface로 설계하신 이유도 궁금하네요 😄

parkjun5 replied:
OAuth2ClientRegistration 과 같은 이유입니다.

parkeeseul commented on src/main/java/nextstep/security/authentication/filter/oauth/OAuth2LoginAuthenticationFilter.java:
```diff
@@ -2,105 +2,128 @@
-        String code = request.getParameter("code");
-        if (StringUtils.hasText(code)) {
-            return code;
+    private Authentication attemptAuthentication(HttpServletRequest request) {
```
현재 코드에서는 OAuth2 인증을 수행할 때 `UserDetailsService`를 사용하여 `UserDetails` 기반으로 인증하고 있는데요,  
Spring Security의 OAuth2 인증 흐름에서는 `OAuth2UserService`를 활용하여 `OAuth2User` 객체를 생성하는 방식이 더 적절할 것 같습니다.  

`OAuth2User`는 OAuth2 제공자로부터 받은 사용자 정보(`attributes`)를 더 유연하게 저장할 수 있고,  
Spring Security의 `OAuth2LoginAuthenticationProvider`에서도 `OAuth2UserService`를 통해 `OAuth2User`를 생성합니다.  
최종적으로 `OAuth2AuthenticationToken`은 `OAuth2User`를 기반으로 만들어지기 때문에,  `UserDetailsService`를 사용하여 인증하는 방식보다는,  
OAuth2 제공자로부터 받은 사용자 정보를 `OAuth2User` 객체에 저장하고,
OAuth2AuthenticationToken을 생성하는 것이 더 적절할 것 같아요. 😄  

또한, 현재 `OAuth2LoginAuthenticationFilter`에서 OAuth2 인증 처리를 하고 있는데,  
Spring Security에서는 `AuthenticationManager`를 통해 `OAuth2LoginAuthenticationProvider`가 인증을 수행합니다.  

여유가 되신다면, **OAuth2LoginAuthenticationProvider의 내부 구현과 동작 방식**을 함께 고민해 보시면  
OAuth2 인증 흐름을 더 깊이 이해하는 데 도움이 될 것 같습니다! 🚀

parkjun5 replied:
UserDetails 보다는 OAuth2에 맞는 값을 쓰는게 맞겠네요. 내부에 Map을 사용하는 것도 저 마다 다른  OAuth2 값을 유연하게 처리하기 위함일 것 같습니다.
개인적으로 Map은 어떤 값이 전달 될 지 몰라서 조금 선호하지 않지만, default method 를 통해서 정해놓았으니 괜찮은 것 같습니다.

parkeeseul reviewed (changes requested):
2단계 미션도 잘 진행해 주셨네요! 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다.
화이팅 입니다 🔥

parkjun5 committed:
ref: OAuth2AuthenticationRequestResolver HttpServletRequest 를 처리하도록 변경

 - OAuth2 요청 처리를 한 곳에서 관리하도록 책임이 분리

parkjun5 committed:
ref: OAuth2AuthenticationRequestResolver HttpServletRequest 를 처리하도록 변경

 - OAuth2 요청 처리를 한 곳에서 관리하도록 책임이 분리

parkjun5 committed:
feat: Scope String to Set<String> 변환 구현

parkjun5 committed:
refactor: OAuth2 인증은 UserDetails가 아닌 OAuth2User를 사용하도록 변경

 - OAuth2User에는 Map으로 선언된 attributes를 통해 다양한 값이 들어갈 수 있음
 - 다양한 OAuth2 서비스마다 다른 값을 전달할 수 있으므로 이를 유연하게 처리 가능

parkjun5 committed:
polishing

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요 👍 
소소한 코멘트 하나 남겨두었으니 확인 부탁드리고 이만 머지하도록 하겠습니다!
고생하셨습니다! 남은 미션도 화이팅 이예요! 🔥
